server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -829,26 +838,27 @@ public String[] indexAliases(
         Predicate<AliasMetadata> requiredAlias,
         Predicate<DataStreamAlias> requiredDataStreamAlias,
         boolean skipIdentity,
-        Set<String> resolvedExpressions
+        Set<ResolvedExpression> resolvedExpressions
     ) {
-        if (isAllIndices(resolvedExpressions)) {
+        // TODO: This conditional seems to never be true in production code? We resolve expressions first which will always expand _all
+        if (isAllIndicesExpression(resolvedExpressions)) {
             return null;
         }
-
         final IndexMetadata indexMetadata = state.metadata().getIndices().get(index);
         if (indexMetadata == null) {
             // Shouldn't happen
             throw new IndexNotFoundException(index);
         }
 
-        if (skipIdentity == false && resolvedExpressions.contains(index)) {
+        if (skipIdentity == false && resolvedExpressions.contains(new ResolvedExpression(index))) {
             return null;
         }
 
         IndexAbstraction ia = state.metadata().getIndicesLookup().get(index);
         DataStream dataStream = ia.getParentDataStream();
         if (dataStream != null) {
-            if (skipIdentity == false && resolvedExpressions.contains(dataStream.getName())) {
+            // TODO: This conditional also seems to never be true in production code? Data stream names are always expanded to their indices
+            if (skipIdentity == false && resolvedExpressions.contains(new ResolvedExpression(dataStream.getName()))) {
```

Comment:
I also double checked and all the calls happen after the resources have been resolved with `preserveDataStreams: false`. However, this method cannot check this and it is easy to miss :( . Should we also add it as an assertion?

We can always add it in the comments but I see that many the expression resolution might not happen at the direct caller, so it would make sense. What do you think?

Another option is see this as a smell and try to improve this in a subsequent PR. I might actually try to understand it a bit better, for educational purposes.

Reply:
I think I'll leave these in, since as you say there's no way to ensure callers have resolved the expressions in a specific way before this method. I just needed to make sure I wasn't losing it.